File: 백준/Gold/2306. 유전자/유전자.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {

    static String code;
    static int[][] dp;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        code = br.readLine();
        dp = new int[code.length()][code.length()];

        for (int size = 1; size < code.length(); size++) {
            for (int start = 0; start + size < code.length(); start++) {
                int end = start + size;
                
                if ((code.charAt(start) == 'a' && code.charAt(end) == 't')
                        || code.charAt(start) == 'g' && code.charAt(end) == 'c') {
                    dp[start][end] = dp[start + 1][end - 1] + 2;
                }

                for (int mid = start; mid < end; mid++) {
                    dp[start][end] = Math.max(dp[start][end], dp[start][mid] + dp[mid + 1][end]);
                }
            }
        }

        int answer = dp[0][code.length() - 1];

        System.out.println(answer);

    }
}

File: 백준/Gold/2306. 유전자/유전자.py
import sys

input = sys.stdin.readline


code = input().strip()
n = len(code)
dp = [[0] * n for _ in range(n)]

for length in range(1, n):
    for start in range(n):
        end = start + length

        if n <= end:
            break

        if (code[start] == "a" and code[end] == "t") or (code[start] == "g" and code[end] == "c"):
            dp[start][end] = dp[start + 1][end - 1] + 2

        for mid in range(start, end):
            dp[start][end] = max(dp[start][end], dp[start][mid] + dp[mid + 1][end])

print(dp[0][-1])